### PR TITLE
Fix for LLVM crashes

### LINKED ISSUE #1030: Blaze can trigger crashes in the LLVM
```python
import numpy as np
 
from odo import discover
from dask.array import from_array
from dask.array.blaze import compute, symbol
 
nx = np.arange(600).reshape((20, 30))
dx = from_array(nx, blockshape=(4, 5))
sx = symbol('x', discover(dx))
 
expr = sx.mean(axis=0)
 
dask_ns = {sx: dx}
 
result = compute(expr, dask_ns)
result2 = np.array(result)
print(result2)
```
We get the output
```bash
python crash_llvm.py 
python: LeaksContext.h:54: void llvm::LeakDetectorImpl<T>::addGarbage(const T*) [with T = llvm::Value]: Assertion `Ts.count(o) == 0 && "Object already in set!"' failed.
[1]    18320 abort      python test.py
```

### PR DESCRIPTION
**Pull request by cowlicks**
This closes #1030 thanks for the comments from @gmarkall

### CODE REVIEW

File: blaze/compute/numba.py [hunk 1]
```diff
@@ -1,14 +1,15 @@
 from __future__ import absolute_import, division, print_function
 
 import numpy as np
+from toolz import memoize
+import datashape
+import numba
+from numba.npufunc.ufuncbuilder import UFuncBuilder
```
Comment (gmarkall) [inline blaze/compute/numba.py]: I think this might be a typo - `npyufunc` instead?

Comment (cowlicks) [inline blaze/compute/numba.py]: Oh yeah, so this was not even run when I tested it >___<

**Comment by cpcloud** [comment]
Can you add a test for this?

**Comment by mrocklin** [comment]
This is kind of a PITA to test.  The error arose when two threads ran the function before compilation, triggering concurrent JIT compilations.  We'll get indirect coverage from the dask PR when it's in.

**Comment by cpcloud** [comment]
ok, fair enough

**Comment by cpcloud** [comment]
@mrocklin can you confirm that this fixes the issue you're having in dask?

**Comment by cowlicks** [comment]
Okay this does it for me. Tested locally:
```bash
(blaze-dev)➜  ~  python crash_llvm.py
[ 285.  286.  287.  288.  289.  290.  291.  292.  293.  294.  295.  296.
  297.  298.  299.  300.  301.  302.  303.  304.  305.  306.  307.  308.
  309.  310.  311.  312.  313.  314.]
```

**Comment by mrocklin** [comment]
@cowlicks does #1029 pass with this fix?

**Comment by cowlicks** [comment]
It does locally, I rebased #1029 on top of this so y'all can see it works with travis (hopefully it will). After this is in I'll rebase #1029 on master again.

File: blaze/compute/numba.py [hunk 2]
```diff
@@ -196,15 +198,20 @@ def _get_numba_ufunc(expr):
...
+    # get the func
     func = eval(s, scope)
+    # get the signature
     sig = compute_signature(expr)
-    return numba.vectorize([sig], nopython=True)(func)
+    # vectorize is currently not thread safe. So lock the thread.
+    # TODO FIXME remove this when numba has made vectorize thread safe.
+    lock.acquire()
+    ufunc = numba.vectorize([sig], nopython=True)(func)
+    lock.release()
```
Comment (mrocklin) [inline blaze/compute/numba.py]: Just PSA (not a request to change anything) but are you aware that locks function as context managers?

```python
with lock:
    ufunc = ...
```

**Comment by mrocklin** [comment]
@cpcloud ok to merge after 1029 passes tests?

File: blaze/compute/numba.py (hunk 2 above)
Comment (cpcloud) [inline blaze/compute/numba.py]: What would happen if `numba.vectorize` raises? Would the lock be held indefinitely? If so, then this should definitely use the context manager functionality.

**Comment by cpcloud** [comment]
Yep, modulo my comment on the context manager issue

**Comment by mrocklin** [comment]
Looks like this hangs

**Comment by cowlicks** [comment]
Yeah, I've been working on it. Still have not figured it out.

**Comment by cpcloud** [comment]
This might be sqlalchemy related.

**Comment by cpcloud** [comment]
I'm looking into it.

**Comment by cowlicks** [comment]
I think so too. When I remove the sql source from `test_comprehensive` it doesn't hang. sparksql stuff fails. And it hangs later on `test_sql_compute.py::test_computation_directly_on_sqlalchemy_tables`

**Comment by cowlicks** [comment]
@cpcloud What changed recently with sqlalchemy? IsIn stuff?

**Comment by cowlicks** [comment]
Oh #1040 is hanging too. Lol I thought this had something to do with the thread lock never releasing this whole time.

**Comment by cpcloud** [comment]
Sqlalchemy recently released 1.0.0

**Comment by cpcloud** [comment]
Not unheard of to have breaking changes in a 1.0 release :)

**Comment by cpcloud** [comment]
@cowlicks i think you need to pull from master, rehash and push:

```sh
git pull upstream master && git commit --amend -C HEAD && git push -f
```

**Comment by cpcloud** [comment]
actually no need to rehash

**Comment by cpcloud** [comment]
closed by #1029

**Pull request closed without merging**